launch/launch/launch_service.py

```diff
@@ -338,28 +339,34 @@ def _on_exception(loop, context):
                         ret = await self._shutdown(reason='idle', due_to_sigint=False)
                         assert ret is None, ret
                         continue
-                    process_one_event_task = this_loop.create_task(self._process_one_event())
-                    if self.__shutting_down:
-                        # If shutting down and idle then we're done.
-                        if is_idle:
+
+                    # Stop running if we're shutting down and there's no more work
+                    if self.__shutting_down and is_idle:
+                        if (
+                            process_one_event_task is not None and
+                            not process_one_event_task.done()
+                        ):
                             process_one_event_task.cancel()
-                            break
-                        entity_futures = [pair[1] for pair in self._entity_future_pairs]
-                        entity_futures.append(process_one_event_task)
-                        entity_futures.extend(self.__context._completion_futures)
-                        done = set()  # type: Set[asyncio.Future]
-                        while not done:
-                            done, pending = await asyncio.wait(
-                                entity_futures,
-                                timeout=1.0,
-                                return_when=asyncio.FIRST_COMPLETED
-                            )
-                            if not done:
-                                self.__logger.debug(
-                                    'still waiting on futures: {}'.format(entity_futures)
-                                )
-                    else:
-                        await process_one_event_task
+                        break
+
+                    # Collect futures to wait on
+                    entity_futures = [pair[1] for pair in self._entity_future_pairs]
+                    entity_futures.extend(self.__context._completion_futures)
```

Comment:
this is only another option, and not required:
```suggestion
                    entity_futures = []
                    if self.__context._event_queue.empty():
                        entity_futures = [pair[1] for pair in self._entity_future_pairs]
                        entity_futures.extend(self.__context._completion_futures)
```
this will avoid the checking in L358 about when creating a new task for processing an event, and will avoid the previous race.

Maybe it makes the code harder to understand, I don't mind strongly

Reply:
I don't understand how it avoids the condition for creating a new task below. Or are you suggesting that makes it so we only have to wait for the task in the queue at L365, and it doesn't get spuriously woken up by another future completing?

Reply:
> Or are you suggesting that makes it so we only have to wait for the task in the queue at L365, and it doesn't get spuriously woken up by another future completing?

That's exactly my idea, avoiding the spurious waits.
We only need those wakes when the queue is empty, if not we can be waiting for something to be put in the queue when nobody is going to do that.

> I don't understand how it avoids the condition for creating a new task below

Maybe, we still need that check.
Because I imagine that some of the completion futures can generate a new event when they are done.

---

The current code is fine. It avoids the hung, which is the important issue.

Reply:
I've applied your suggestion in fc4e9ef